### ISSUE TITLE
Find replacement weather data API

### ISSUE THREAD

**Issue by alex-hhh** [alex-hhh/ActivityLog2#46]
## TLDR

The weather API provider used by ActivityLog2 [will discontinue](https://blog.darksky.net/) their public weather API and we need to find another way to retrieve weather information for imported activities. 

Also, this happened before: #33

## Update (April 2022)

Weather records stored in FIT files can be imported and displayed by the GUI, reducing the need for a web service for those who have devices that record this data (I could not find how to get my FR945 to record it though...).  The implementation has the following limitations:

* The FIT reader can read "current", "forecast" and "warning" weather records, but only "current" ones are stored in the database.
* Multiple weather records can be present in a FIT file and they are all stored in the database associated with the session, but only the first one is displayed in the GUI.
* When weather records are present in the FIT file, the application will no longer request weather data from the web service and also editing this data is not possible using the "Activity/Edit Weather..." menu.
* FIT weather records don't have Dew Point, so Humidex cannot be calculated.

## Previous Situation (September 2021)

DarkSky will continue to support existing API keys until the end of 2022, so the current binary release will continue to have weather data until than.

Several other weather services have appeared aiming to replace DarkSky (seemed to be a popular service).  Some of these are linked in comments to this issue, to be investigated once I start working on it.

## Previous situation (April 2020)

At this time, DS stopped providing new API keys, but existing ones will work until the end of 2021 -- this is a vague date, but their [blog post](https://blog.darksky.net/) does not have more details.

The current binary release is shipped with an API key and this will continue to work, perhaps for another 12 months or so, but anyone who wants to build their own package won't be able to get weather data.

## What next?

* **Find a new Weather API**  Unfortunately a quick web search did not find a suitable one (see Weather API requirements below). 
* **DONE** -- **Manual Weather Data Entry** ActivityLog2 allows entering weather data manually and services such as Weather Underground still allow retrieving the historical weather information using their Web interface.  This is not ideal as it involves a lot of manual work.
* **DONE** -- **Weather Records in FIT Files** some Garmin devices record and store weather data in the activity files (presumably by querying the current weather at the time).  AL2 currently discards this information, but it could be imported as well.  Only a small set of Garmin devices collect this data, so this is is a partial solution at best.
* **Portable Weather Stations**.  These devices seem to exist on the market and perhaps some of them could be used to automatically retrieve data from them
* **Investigate writing our own weather station** this is a "pie in the sky" project, but after all ActivityLog2 is a hobby and might be interesting to experiment with this.  There are plenty of Arduino weather station projects on the web.

## Weather API Requirements

To make use of a Weather API by activity log, the API needs to provide the following features:

* **Lookup weather based on GPS location**.  ActivityLog2 retrieves weather data for activities automatically and this is done by looking at the GPS location for an activity and querying the Weather API
* **Provide Historical Data**.  Activities are usually imported a few hours to a few days after they were recorded, and weather data is retrieved at the time of import.  This means the Weather API must provide the ability to query the weather at some time in the recent past.  As a bonus, it would be nice to be able to retrieve weather several years in the past for people who import their data for the first time.
* **Provide a Free (no cost) Access Tier**   ActivityLog2 is a hobby project, and I don't make any money off it.  I cannot justify paying a monthly subscription for this service.  Plus, accessing historical data can cost a lot more.  For example, OpenWeatherMap provides a free tier, but this only includes access to current and forecast weather and the cheapest plan which provides historical weather data starts at $150/month.
* **Allow storing weather data locally** The Terms of Service (TOS) for the service must allow storing weather data locally along with the session data, so ActivityLog2 does not have to fetch weather info each time the user looks at an old activity
* **Allow displaying additional weather info** The Terms of Service (TOS) for the service must allow showing additional weather information along with the rest of the weather data.  ActivityLog2 shows temperature recorded by the device if this is available in the FIT file as well as a computed "Humidex" value.

**Comment by alex-hhh** [comment]
Something to investigate later: https://www.climacell.co/blog/top-8-weather-apis-for-2020/

**Comment by alex-hhh** [comment]
See also: https://meteostat.net/en and their API: https://api.meteostat.net

**Comment by alex-hhh** [comment]
Also see [this reddit thread](https://www.reddit.com/r/programming/comments/nbepec/for_anyone_using_the_soon_to_shut_down_dark_sky/) and related web site: https://pirateweather.net/ .

**Comment by alex-hhh** [comment]
Another one: https://news.ycombinator.com/item?id=28499910

**Comment by open-meteo** [comment]
Creator of open-meteo here

Let me know if I can be of help. Do you have additional requirements. 

For which weather variables do you need historical data?

Best, Jeff

**Comment by alex-hhh** [comment]
Hi Jeff,

ActivityLog2 currently stores the following weather related values: temperature, dew point, relative humidity, wind (speed, gusts and direction) and barometric pressure.

However, this issue is currently on hold for me, Dark Sky have extended their availability until the end of 2022 and I have limited time to work on this project an currently focusing on other things.  I will probably revisit this issue sometimes next year.

Cheers,
Alex.

**Comment by alex-hhh** [comment]
Looked at **AccuWeather** API and found it unsuitable: 

* Their terms of service (TOS) explicitly prohibit storing the retrieved weather data into a database -- given that their API can retrieve weather data for up to 24 hours in the past, this makes it impossible to show weather data for old activities (AL2 can store weather data in the database, but this is prohibited by AccuWeather TOS)
* Also, their TOS don't permit showing their weather data next to other weather info (AL2 shows temperature recorded in the FIT file.

**Comment by alex-hhh** [comment]
Removed automatic weather data download from ActivityLog2, since no suitable service was found (See criteria above).  ActivityLog2 uses SQLite to store the data, and it is easy (for technical people anyway) to write a script that automatically downloads data from any weather provider and updates it in the database.  

A sample script to fetch data from Meteostat is provided here: https://gist.github.com/alex-hhh/e1a4ffe3e861ad3fd31f6655a3a66b26

**Comment by nickb937** [comment]
I think the new One Call API 3.0 from OpenWeatherMap might be suitable now https://openweathermap.org/full-price#current because it includes current and historical weather information.

Each ActivityLog2 user would need to apply for their own API keys and you'd probably need to setup a way to stop the number of API calls exceeding 1000 in a given 24 hour period.

**Comment by alex-hhh** [comment]
Hi @nickb937 , looking at the link you provided, it looks like the situation with Open Weather Map is the same as when I last looked at it: historical weather download (1 month back) is only available in the "Starter" category at 120 GBP / month ($150 USD / month) -- there is no free tier access to this data.

As explained in the *Weather API Requirements* section above, ActivityLog2 must be able to query the weather as it was a few hours / days ago when the run, cycle or other activity was completed -- this is covered by historical weather data for most weather APIs.

For now, I have been using data from Meteostat using the script provided here: https://gist.github.com/alex-hhh/e1a4ffe3e861ad3fd31f6655a3a66b26, and seems to be working OK, I might look at adding direct support for Meteostat in ActivityLog2, but I won't do it in the near term, as I a busy with other things.    You can also modify that script to fetch data from OpenWeatherMap (or any other API provider), if you wish to experiment with them.
